cdap-app-fabric/src/test/java/co/cask/cdap/reporting/ProgramHeartBeatDatasetTest.java

```diff
@@ -0,0 +1,246 @@
+                                      TransactionExecutor txnl, ProgramHeartbeatDataset programHeartbeatDataset)
+    throws InterruptedException, TransactionFailureException {
+    txnl.execute(() -> {
+      for (long time = startTime + interval; time < endTime; time += interval) {
+        programHeartbeatDataset.writeRunRecordMeta(runRecordMeta, time);
+      }
+    });
+  }
+
+  @Test
+  public void testScanningProgramStatus() throws Exception {
+    // p1 : x           ->    x + 5
+    // scanning from x -> x + 4 -> return one record with status running
+    // scanning from x -> x + 5min + 1s -> return one record with status killed
+
+    ProgramHeartbeatDataset programHeartbeatDataset = getHeartBeatStore("testStatusChange");
+    TransactionExecutor txnl = getTxExecutor(programHeartbeatDataset);
+
+    //  write program status "running" for program 1 starting at x
+    long startTime1 = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
+    RunId runId = RunIds.generate();
+    RunRecordMeta meta =
+      getMockRunRecordMeta(NamespaceId.DEFAULT,
+                           runId, new StatusTime(ProgramRunStatus.RUNNING, startTime1), null);
+    txnl.execute(() -> {
+      programHeartbeatDataset.writeRunRecordMeta(meta, startTime1);
+    });
+
+    // write heart beat messages for 10 minutes (every minute) for this program run.
+    long heartBeatEndTime = startTime1 + TimeUnit.MINUTES.toSeconds(4);
+    long interval = TimeUnit.MINUTES.toSeconds(1);
+    setUpProgramHeartBeats(meta, startTime1, heartBeatEndTime, interval, txnl, programHeartbeatDataset);
+
+    long programEndTime = startTime1 + TimeUnit.MINUTES.toSeconds(5);
+
+    RunRecordMeta metaKilled =
+      getMockRunRecordMeta(NamespaceId.DEFAULT,
+                           runId, new StatusTime(ProgramRunStatus.KILLED, programEndTime), null);
+    txnl.execute(() -> {
+      programHeartbeatDataset.writeRunRecordMeta(metaKilled, programEndTime);
+    });
+
+    // perform scan checks
+    txnl.execute(() -> {
+      byte[] startRowKey1 = Bytes.toBytes(startTime1);
+      byte[] stillRunning = Bytes.toBytes(heartBeatEndTime);
+      byte[] endRowKey = Bytes.toBytes(programEndTime + 1);
+      // end row key is exclusive, scanning from x : x + 5 should only return run1
+      Collection<RunRecordMeta> runRecordMetaList = programHeartbeatDataset.scan(startRowKey1, stillRunning);
+      Assert.assertEquals(1 , runRecordMetaList.size());
+      Assert.assertEquals(ProgramRunStatus.RUNNING, runRecordMetaList.iterator().next().getStatus());
+      // scanning from x : x + (5 min + 1s), should return run1 with status KILLED
+      runRecordMetaList = programHeartbeatDataset.scan(startRowKey1, endRowKey);
+      Assert.assertEquals(1 , runRecordMetaList.size());
+      Assert.assertEquals(ProgramRunStatus.KILLED, runRecordMetaList.iterator().next().getStatus());
+    });
+  }
+
+  private TransactionExecutor getTxExecutor(ProgramHeartbeatDataset programHeartbeatDataset) {
```

Comment:
in general, we should avoid methods that are just a single line that calls another method. It adds an extra layer of indirection without real benefit.